### PR TITLE
Sprint1

### PR DESCRIPTION
**Pull request by BenoitManhes**
- Create the Application
- Create repository to download data : Flow not used for LiveData just now
- Display Data in a RecyclerView

### CODE REVIEW

File: app/src/main/java/com/example/lunabeeusers/ui/overview/UserOverviewFragment.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import android.view.View
+import android.view.ViewGroup
+import com.example.lunabeeusers.R
+
+import com.example.lunabeeusers.databinding.UserOverviewFragmentBinding
+
+class UserOverviewFragment : Fragment() {
+
+    private val viewModel: UserOverviewViewModel by lazy {
+        ViewModelProviders.of(this).get(UserOverviewViewModel::class.java)
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/ui/overview/UserOverviewFragment.kt:16]: Deprecated

File: app/src/main/java/com/example/lunabeeusers/ui/overview/UserListAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+         * Setting space between Recycler View’s cell
+         */
+        override fun getItemOffsets(outRect: Rect, view: View,
+                                    parent: RecyclerView, state: RecyclerView.State) {
+            with(outRect) {
+                if (parent.getChildAdapterPosition(view) == 0) {
+                    top = spaceHeight
+                }
+//                left =  spaceHeight
+//                right = spaceHeight
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/ui/overview/UserListAdapter.kt]: Don't leave commented code

File: app/build.gradle [hunk 3]
```diff
@@ -23,15 +30,52 @@ android {
...
+    }
 }
 
 dependencies {
     implementation fileTree(dir: 'libs', include: ['*.jar'])
-    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
-    implementation 'androidx.appcompat:appcompat:1.1.0'
-    implementation 'androidx.core:core-ktx:1.3.1'
-    implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
+    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
```
Comment (qmetzler-luna) [inline app/build.gradle:45]: First two imports are useless

File: build.gradle [hunk 4]
```diff
@@ -1,15 +1,33 @@
...
 buildscript {
-    ext.kotlin_version = '1.3.71'
+
+    ext {
+        // Versions for all the dependencies we plan to use.
+        version_android_gradle_plugin = "3.6.3"
+        version_core = "1.3.1"
+        version_constraint_layout = "1.1.3"
+        version_glide = "4.11.0"
+        version_kotlin = "1.3.71"
```
Comment (qmetzler-luna) [inline build.gradle]: latest version is 1.3.72

File: build.gradle [hunk 5]
```diff
@@ -1,15 +1,33 @@
...
+        version_kotlin_coroutines = "1.1.0"
+        version_lifecycle_extensions = "2.0.0"
+        version_moshi = "1.8.0"
+        version_navigation = "2.3.0"
+        version_retrofit = "2.5.0"
+        version_retrofit_coroutines_adapter = "0.9.2"
```
Comment (qmetzler-luna) [inline build.gradle:16]: The latest version is 2.9.0 + it already contains a way to use coroutines without importing another lib

File: build.gradle [hunk 6]
```diff
@@ -1,15 +1,33 @@
...
-    ext.kotlin_version = '1.3.71'
+
+    ext {
+        // Versions for all the dependencies we plan to use.
+        version_android_gradle_plugin = "3.6.3"
+        version_core = "1.3.1"
+        version_constraint_layout = "1.1.3"
+        version_glide = "4.11.0"
+        version_kotlin = "1.3.71"
+        version_kotlin_coroutines = "1.1.0"
```
Comment (qmetzler-luna) [inline build.gradle]: The latest version is 1.3.8

File: app/build.gradle [hunk 7]
```diff
@@ -23,15 +30,52 @@ android {
...
+    // ViewModel and LiveData
+    implementation "androidx.lifecycle:lifecycle-extensions:$version_lifecycle_extensions"
+
+    // Navigation
+    implementation "androidx.navigation:navigation-fragment-ktx:$version_navigation"
+    implementation "androidx.navigation:navigation-ui-ktx:$version_navigation"
+
+    // Moshi (Convert JSON to Object)
+    implementation "com.squareup.moshi:moshi:$version_moshi"
+    implementation "com.squareup.moshi:moshi-kotlin:$version_moshi"
```
Comment (qmetzler-luna) [inline app/build.gradle:66]: Useless to import moshi it's already contained in com.squareup.retrofit2:converter-moshi
You can use `MoshiConverterFactory.create()` straight in ApiService.kt and replace `@Json` by `@field:Json` in User.kt

File: app/src/main/java/com/example/lunabeeusers/data/network/ApiService.kt [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+// Moshi Builder
+// Used to convert a JSON to a specific object
+private val moshi = Moshi.Builder()
+    .add(KotlinJsonAdapterFactory())
+    .build()
+
+// Retrofit Builder
+private val retrofit = Retrofit.Builder()
+    .addConverterFactory(MoshiConverterFactory.create(moshi))
+    .addCallAdapterFactory(CoroutineCallAdapterFactory())
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/data/network/ApiService.kt:23]: Should be useless

File: app/src/main/java/com/example/lunabeeusers/data/network/ApiService.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    .baseUrl(BASE_URL)
+    .build()
+
+interface ApiService {
+
+    @GET("users")
+    fun getUsers():
+            Deferred<List<User>>
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/data/network/ApiService.kt]: Should be something more like
```
@GET("users")
suspend fun getUsers(): List<User>
```

File: app/src/main/java/com/example/lunabeeusers/ui/overview/UserOverviewViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    // Coroutines
+    private var viewModelJob = Job()
+    private var coroutineScope = CoroutineScope(viewModelJob + Dispatchers.Main)
+
+    init{
+        getUsersFromApi()
+    }
+
+    private fun getUsersFromApi(){
+        var getUsersDeferred = UsersApi.retrofitService.getUsers()
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/ui/overview/UserOverviewViewModel.kt:32]: Can be val

File: app/src/main/java/com/example/lunabeeusers/ui/overview/UserOverviewViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+        coroutineScope.launch {
+            try {
+                var listResult = getUsersDeferred.await()
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/ui/overview/UserOverviewViewModel.kt:36]: Can be val

File: app/src/main/java/com/example/lunabeeusers/ui/overview/UserOverviewViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    // The internal MutableLiveData List<User> that stores the users return by the API
+    private val _usersList = MutableLiveData<List<User>>()
+
+    // External MutableLiveData List<User> used to diplay users
+    val userList: LiveData<List<User>>
+        get() = _usersList
+
+    // Coroutines
+    private var viewModelJob = Job()
+    private var coroutineScope = CoroutineScope(viewModelJob + Dispatchers.Main)
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/ui/overview/UserOverviewViewModel.kt:25]: You can use **viewModelScope** imported by _lifecycle-viewmodel-ktx_ to remove some boiler plate code

File: app/src/main/res/layout/user_overview_fragment.xml [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        tools:context="com.example.android.marsrealestate.MainActivity">
+
+        <androidx.recyclerview.widget.RecyclerView
+            android:id="@+id/users_rv"
+            android:layout_width="0dp"
+            android:layout_height="0dp"
+            android:clipToPadding="false"
+            android:padding="6dp"
```
Comment (qmetzler-luna) [inline app/src/main/res/layout/user_overview_fragment.xml:22]: Dimens should never be hardcoded and always be taken from the dimens.xml file

File: app/src/main/java/com/example/lunabeeusers/ui/overview/UserOverviewFragment.kt [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    override fun onCreateView( inflater: LayoutInflater, container: ViewGroup?,
+                               savedInstanceState: Bundle? ): View? {
+
+        val binding = UserOverviewFragmentBinding.inflate(inflater)
+
+        // Allows Data Binding to Observe LiveData with the lifecycle of this Fragment
+        binding.setLifecycleOwner(this)
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/ui/overview/UserOverviewFragment.kt:25]:
```suggestion
        binding.lifecycleOwner = this
```

File: app/src/main/res/layout/item_user_list.xml [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+            app:layout_constraintTop_toTopOf="parent"
+            app:layout_constraintStart_toEndOf="@+id/userAvatarIv"
+            app:layout_constraintEnd_toEndOf="parent"
+            >
+
+
+            <TextView
+                android:id="@+id/firstnameTv"
+                android:layout_width="wrap_content"
+                app:firstname="@{user}"
```
Comment (qmetzler-luna) [inline app/src/main/res/layout/item_user_list.xml:51]: Why are you not using user.firstname?

File: app/src/main/res/layout/item_user_list.xml [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+                style="@style/textview_user_item"
+                tools:text="Woody" />
+
+            <TextView
+                android:id="@+id/lastnameTv"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_marginStart="@dimen/short_margin"
+                app:lastname="@{user}"
```
Comment (qmetzler-luna) [inline app/src/main/res/layout/item_user_list.xml:60]: Why are you not using user.lastname?

File: app/src/main/java/com/example/lunabeeusers/core/BindingAdapter.kt [hunk 17]
```diff
@@ -0,0 +1,42 @@
+package com.example.lunabeeusers.core
+
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.core.net.toUri
+import androidx.databinding.BindingAdapter
+import androidx.recyclerview.widget.RecyclerView
+import com.bumptech.glide.Glide
+import com.bumptech.glide.request.RequestOptions
+import com.example.lunabeeusers.R
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/core/BindingAdapter.kt:10]: Unused imports

File: app/src/main/java/com/example/lunabeeusers/core/BindingAdapter.kt [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    }
+}
+
+@BindingAdapter("imageUrl")
+fun bindImage(imgView: ImageView, imgUrl: String?) {
+    imgUrl?.let {
+        val imgUri = imgUrl.toUri().buildUpon().scheme("https").build()
+        Glide.with(imgView.context)
+            .load(imgUri)
+            .into(imgView)
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/core/BindingAdapter.kt:40]: No need for an imgUri. imgUrl can be use directly.

File: app/src/main/java/com/example/lunabeeusers/data/network/ApiService.kt [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+import com.example.lunabeeusers.data.model.User
+import com.jakewharton.retrofit2.adapter.kotlin.coroutines.CoroutineCallAdapterFactory
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
+import kotlinx.coroutines.Deferred
+import retrofit2.Retrofit
+import retrofit2.converter.moshi.MoshiConverterFactory
+import retrofit2.http.GET
+
+private const val BASE_URL = "http://server.lunabee.studio:11111/techtest/"
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/data/network/ApiService.kt:12]: Could be in UsersApi

File: app/src/main/java/com/example/lunabeeusers/data/network/ApiService.kt [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+private val moshi = Moshi.Builder()
+    .add(KotlinJsonAdapterFactory())
+    .build()
+
+// Retrofit Builder
+private val retrofit = Retrofit.Builder()
+    .addConverterFactory(MoshiConverterFactory.create(moshi))
+    .addCallAdapterFactory(CoroutineCallAdapterFactory())
+    .baseUrl(BASE_URL)
+    .build()
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/data/network/ApiService.kt:25]: Could be in UsersApi

File: app/src/main/java/com/example/lunabeeusers/data/network/ApiService.kt [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+    @GET("users")
+    fun getUsers():
+            Deferred<List<User>>
+}
+
+object UsersApi{
+    val retrofitService: ApiService by lazy {
+        retrofit.create(ApiService::class.java)
+    }
+}
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/data/network/ApiService.kt:38]: Try to have only one class by file

File: app/src/main/java/com/example/lunabeeusers/ui/overview/UserListAdapter.kt [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+                return UserViewHolder(binding)
+            }
+        }
+    }
+
+    /**
+     * Use to optimize view displaying
+     * When data change, allow to update only view related to data updtated
+     */
+    class DiffCallback : DiffUtil.ItemCallback<User>() {
```
Comment (qmetzler-luna) [inline app/src/main/java/com/example/lunabeeusers/ui/overview/UserListAdapter.kt:45]: Can be private

File: app/src/main/res/values/colors.xml [hunk 23]
```diff
@@ -3,4 +3,9 @@
     <color name="colorPrimary">#6200EE</color>
     <color name="colorPrimaryDark">#3700B3</color>
     <color name="colorAccent">#03DAC5</color>
+
+    <color name="white">#FFFFFF</color>
```
Comment (qmetzler-luna) [inline app/src/main/res/values/colors.xml]: android:color/white already exists

File: app/src/main/res/values/styles.xml [hunk 24]
```diff
@@ -8,4 +8,12 @@
         <item name="colorAccent">@color/colorAccent</item>
     </style>
 
+    <style name="textview_user_item">
+        <item name="android:layout_width">wrap_content</item>
+        <item name="android:layout_height">wrap_content</item>
```
Comment (qmetzler-luna) [inline app/src/main/res/values/styles.xml]: android:layout_width and android:layout_height shouldn't be in the style

File: app/src/main/res/values/styles.xml [hunk 25]
```diff
@@ -8,4 +8,12 @@
...
+        <item name="android:fontFamily">@font/heebo_light</item>
+        <item name="android:textColor">@color/purple_dark</item>
+        <item name="android:textSize">18sp</item>
```
Comment (qmetzler-luna) [inline app/src/main/res/values/styles.xml]: should be in the dimens

File: app/src/main/res/values/styles.xml [hunk 26]
```diff
@@ -8,4 +8,12 @@
         <item name="colorAccent">@color/colorAccent</item>
     </style>
 
+    <style name="textview_user_item">
```
Comment (qmetzler-luna) [inline app/src/main/res/values/styles.xml]: style should always inherit from a system one if possible. For example TextAppearance.AppCompat.Body1 here.
naming shoud also follow the convention. For example TextAppearance.App.Body

File: app/src/main/res/layout/item_user_list.xml [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+            app:layout_constraintStart_toEndOf="@+id/userAvatarIv"
+            app:layout_constraintEnd_toEndOf="parent"
+            >
+
+
+            <TextView
+                android:id="@+id/firstnameTv"
+                android:layout_width="wrap_content"
+                app:firstname="@{user}"
+                style="@style/textview_user_item"
```
Comment (qmetzler-luna) [inline app/src/main/res/layout/item_user_list.xml]: should be android:textAppearance

**Review by qmetzler-luna — CHANGES_REQUESTED** [review]
- Coding style is not respected

**Comment by qmetzler-luna** [comment]
@BenoitManhes Tu n'as pas corrigé tous mes retours. C'est parce que tu veux le faire plus tard, parce que c'est pas clair ou un oublie?

**Comment by BenoitManhes** [comment]
@qmetzler-luna C'est bon c'est corrigé. J'ai préféré corriger ton retour concernant retrofit_coroutines_adapter dans le Sprint2 après avoir changer l'architecture et ajouter l'injection

**Pull request merged**